pkg/streamer/reader_test.go

```diff
@@ -0,0 +1,689 @@
+		}
+		obtainExtraEvents2 = append(obtainExtraEvents2, ev)
+		if len(obtainExtraEvents2) == cap(obtainExtraEvents2) {
+			break
+		}
+	}
+	t.verifyNoEventsInStreamer(c, s)
+
+	// verify obtain extra events
+	c.Assert(obtainExtraEvents2, DeepEquals, baseEvents[1:])
+
+	// NOTE: load new UUIDs dynamically not supported yet
+
+	// close the reader
+	err = r.Close()
+	c.Assert(err, IsNil)
+}
+
+func (t *testReaderSuite) TestStartSyncError(c *C) {
+	var (
+		baseDir = c.MkDir()
+		UUIDs   = []string{
+			"b60868af-5a6f-11e9-9ea3-0242ac160006.000001",
+		}
+		cfg      = &BinlogReaderConfig{RelayDir: baseDir}
+		startPos = gmysql.Position{Name: "test-mysql-bin|000001.000001"} // from the first relay log file in the first sub directory
+	)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	r := NewBinlogReader(cfg)
+
+	// no startup pos specified
+	s, err := r.StartSync(gmysql.Position{})
+	c.Assert(errors.Cause(err), Equals, ErrBinlogFileNotSpecified)
+	c.Assert(s, IsNil)
+
+	// empty UUIDs
+	s, err = r.StartSync(startPos)
+	c.Assert(err, IsNil)
+	ev, err := s.GetEvent(ctx)
+	c.Assert(err, ErrorMatches, ".*empty UUIDs not valid.*")
+	c.Assert(ev, IsNil)
+	c.Assert(r.Close(), IsNil)
+
+	// write UUIDs into index file
+	r = NewBinlogReader(cfg) // create a new reader
+	uuidBytes := t.uuidListToBytes(c, UUIDs)
+	err = ioutil.WriteFile(r.indexPath, uuidBytes, 0644)
+	c.Assert(err, IsNil)
+
+	// the startup relay log file not found
+	s, err = r.StartSync(startPos)
+	c.Assert(err, IsNil)
+	ev, err = s.GetEvent(ctx)
+	c.Assert(err, ErrorMatches, fmt.Sprintf(".*%s.*not found.*", startPos.Name))
+
+	// can not re-start the reader
+	s, err = r.StartSync(startPos)
+	c.Assert(errors.Cause(err), Equals, ErrReaderRunning)
+	c.Assert(s, IsNil)
+
+	c.Assert(ev, IsNil)
+	c.Assert(r.Close(), IsNil)
+}
+
+func (t *testReaderSuite) genBinlogEvents(c *C, latestPos uint32) []*replication.BinlogEvent {
+	var (
+		header = &replication.EventHeader{
+			Timestamp: uint32(time.Now().Unix()),
+			ServerID:  11,
+		}
+		events = make([]*replication.BinlogEvent, 0, 10)
+	)
+
+	if latestPos <= 4 { // generate a FormatDescriptionEvent if needed
+		ev, err := event.GenFormatDescriptionEvent(header, 4)
+		c.Assert(err, IsNil)
+		latestPos = ev.Header.LogPos
+		events = append(events, ev)
+	}
+
+	// for these tests, generates some DDL events is enough
+	count := 5 + rand.Intn(5)
+	for i := 0; i < count; i++ {
+		schema := []byte(fmt.Sprintf("db_%d", i))
+		query := []byte(fmt.Sprintf("CREATE TABLE %d (c1 INT)", i))
+		ev, err := event.GenQueryEvent(header, latestPos, 0, 0, 0, nil, schema, query)
+		c.Assert(err, IsNil)
+		latestPos = ev.Header.LogPos
+		events = append(events, ev)
+	}
+
+	return events
+}
+
+func (t *testReaderSuite) purgeStreamer(c *C, s Streamer) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
```

Comment:
it seems that we can set a shorter timeout to save unit test time

Reply:
Yes, done in https://github.com/pingcap/dm/pull/140/commits/ec91143d6bb3dea85021f1f3384bff2a51a02148.